src/jit/codegenxarch.cpp

```diff
@@ -176,23 +176,63 @@ void                CodeGen::genEmitGSCookieCheck(bool pushReg)
 
     // Make sure that EAX is reported as live GC-ref so that any GC that kicks in while
     // executing GS cookie check will not collect the object pointed to by EAX.
-    if (!pushReg && (compiler->info.compRetType == TYP_REF))
-        gcInfo.gcRegGCrefSetCur |= RBM_INTRET;    
+    if (!pushReg)
+    {
+
+#ifdef FEATURE_UNIX_AMD64_STRUCT_PASSING
+
+        if (varTypeIsStruct(compiler->info.compRetType))
+        {
```

Comment:
Why is this if check is needed?

As per my understanding we don't need the above if check and just the below check is sufficient

if (compiler->compMethodReturnsMultiRegRetType())

Note that compMethodReturnsMultiRegRetType() would return true only for struct returned in two registers.  Struct returned in a single register is normalized to byte/short/int/long/ref/byref.  See lvaInitTypeRef() - in that routine we normalize return type and assign it to info.compRetNativeType.

We should be using compRetNativeType on line 216 below.